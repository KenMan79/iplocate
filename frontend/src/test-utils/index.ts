export * from "./test-utils";
export * from "./ipdata";
