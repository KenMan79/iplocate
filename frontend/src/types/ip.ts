export interface IP {
  address: string;
  city: string;
  latitude: number;
  longitude: number;
}
