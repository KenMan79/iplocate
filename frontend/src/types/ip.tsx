export interface IP {
  address: string;
  latitude: string;
  longitude: string;
  city: string;
}
