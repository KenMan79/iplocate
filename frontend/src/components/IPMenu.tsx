import React from "react";
import { FaEye, FaEyeSlash, FaTrash } from "react-icons/fa";
import styled from "styled-components";
import { IP } from "types";

interface Props {
  ips: IP[];
  onSetCurrentIp(ip: IP): void;
  onRemoveIp(ip: IP): void;
  onToggleIpVisible(ip: IP): void;
}

export const IPMenu: React.FC<Props> = (props) => {
  const { ips, onSetCurrentIp, onToggleIpVisible, onRemoveIp } = props;

  if (ips.length === 0) return null;

  return (
    <Wrapper>
      {ips.map((ip) => (
        <Row key={ip.traits.ipAddress}>
          <RowText onClick={() => onSetCurrentIp(ip)}>
            {ip.traits.ipAddress}
          </RowText>
          <div>
            <RowAction
              onClick={() => onRemoveIp(ip)}
              aria-label="remove ip address"
            >
              <FaTrash />
            </RowAction>
            {ip.hidden ? (
              <RowAction
                onClick={() => onToggleIpVisible(ip)}
                aria-label="toggle ip visibility on"
              >
                <FaEye />
              </RowAction>
            ) : (
              <RowAction
                onClick={() => onToggleIpVisible(ip)}
                aria-label="toggle ip visibility off"
              >
                <FaEyeSlash />
              </RowAction>
            )}
          </div>
        </Row>
      ))}
    </Wrapper>
  );
};

const RowAction = styled.button`
  border-radius: 50%;
  flex: 0;
  width: 45px;
  height: 45px;
  font-size: 1.2rem;
  margin: 0 2px;
  display: inline-flex;
  align-items: center;
  justify-content: center;

  :hover {
    background-color: rgba(0, 0, 0, 0.1);
  }
`;

const RowText = styled.button`
  display: block;
  background: none;
  text-align: left;
  font-size: 1.2rem;
  margin: 0;
  font-family: var(--font-family-mono);
  padding: 0.8rem 0;

  :hover {
    text-decoration: underline;
  }
`;

const Row = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
`;

const Wrapper = styled.div`
  padding-top: 1rem;
  color: var(--text-primary);
  button {
    background: none;
    outline: none;
    border: none;
    cursor: pointer;
  }
`;
