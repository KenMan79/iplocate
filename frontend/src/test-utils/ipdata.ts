import { IP } from "types";

export const baseIP: IP = {
  continent: {
    code: "EU",
    geonameId: 6255148,
    names: {
      en: "Europe",
    },
  },
  country: {
    geonameId: 3017382,
    isInEuropeanUnion: true,
    isoCode: "FR",
    names: {
      en: "France",
    },
  },
  traits: {
    isAnonymous: false,
    isAnonymousProxy: false,
    isAnonymousVpn: false,
    isHostingProvider: false,
    isLegitimateProxy: false,
    isPublicProxy: false,
    isResidentialProxy: false,
    isSatelliteProvider: false,
    isTorExitNode: false,
    ipAddress: "2.2.2.2",
    network: "2.2.0.0/18",
  },
  city: {},
  location: {
    accuracyRadius: 500,
    latitude: 48.8582,
    longitude: 2.3387,
    timeZone: "Europe/Paris",
  },
};
