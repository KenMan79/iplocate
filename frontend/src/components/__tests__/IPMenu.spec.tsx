import React from "react";
import { render, baseIP, fireEvent, wait } from "test-utils";
import { IP } from "types";
import { IPMenu } from "../IPMenu";

// ['1.1.1.1', '2.2.2.2', '3.3.3.3']
const allIps: IP[] = Array.from({ length: 3 }).map((_, idx) => ({
  ...baseIP,
  traits: {
    ...baseIP.traits,
    ipAddress: Array.from({ length: 4 })
      .map((_) => idx + 1)
      .join("."),
  },
}));

const setCurrentMock = jest.fn();
const removeMock = jest.fn();
const toggleVisibleMock = jest.fn();

afterEach(() => {
  setCurrentMock.mockReset();
  removeMock.mockReset();
  toggleVisibleMock.mockReset();
});

const renderMocked = (ips: IP[]) => {
  return render(
    <IPMenu
      ips={ips}
      onSetCurrentIp={setCurrentMock}
      onRemoveIp={removeMock}
      onToggleIpVisible={toggleVisibleMock}
    />
  );
};

it("matches snapshot", () => {
  const { asFragment } = renderMocked(allIps);
  expect(asFragment()).toMatchSnapshot();
});

it("sets current ip on address click", async () => {
  const { getByText } = renderMocked(allIps);
  expect(getByText("1.1.1.1")).toBeInTheDocument();
  fireEvent.click(getByText("1.1.1.1"));
  await wait(() => {
    expect(setCurrentMock).toHaveBeenCalledTimes(1);
    expect(setCurrentMock.mock.calls[0][0]).toMatchObject(allIps[0]);
  });
});

it("removes ip address on trash icon click", async () => {
  const { getAllByLabelText } = renderMocked(allIps);
  expect(getAllByLabelText("remove ip address")).toHaveLength(3);
  fireEvent.click(getAllByLabelText("remove ip address")[1]);
  await wait(() => {
    expect(removeMock).toHaveBeenCalledTimes(1);
    expect(removeMock.mock.calls[0][0]).toMatchObject(allIps[1]);
  });
});

it("toggles visibility of ip on eye icon click", async () => {
  const { getAllByLabelText } = renderMocked(allIps);
  expect(getAllByLabelText("toggle ip visibility off")).toHaveLength(3);
  fireEvent.click(getAllByLabelText("toggle ip visibility off")[1]);
  await wait(() => {
    expect(toggleVisibleMock).toHaveBeenCalledTimes(1);
    expect(toggleVisibleMock.mock.calls[0][0]).toMatchObject(allIps[1]);
  });
});

it("returns nothing if there are no ips", async () => {
  const { container } = renderMocked([]);
  expect(container.childElementCount).toEqual(0);
});
