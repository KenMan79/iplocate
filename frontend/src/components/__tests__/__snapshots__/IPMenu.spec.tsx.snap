// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c3 {
  border-radius: 50%;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  width: 45px;
  height: 45px;
  font-size: 1.2rem;
  margin: 0 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3:hover {
  background-color: rgba(0,0,0,0.1);
}

.c2 {
  display: block;
  background: none;
  text-align: left;
  font-size: 1.2rem;
  margin: 0;
  font-family: var(--font-family-mono);
  padding: 0.8rem 0;
}

.c2:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  padding-top: 1rem;
  color: var(--text-primary);
}

.c0 button {
  background: none;
  outline: none;
  border: none;
  cursor: pointer;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        1.1.1.1
      </button>
      <div>
        <button
          aria-label="remove ip address"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </button>
        <button
          aria-label="toggle ip visibility off"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        2.2.2.2
      </button>
      <div>
        <button
          aria-label="remove ip address"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </button>
        <button
          aria-label="toggle ip visibility off"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        3.3.3.3
      </button>
      <div>
        <button
          aria-label="remove ip address"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </button>
        <button
          aria-label="toggle ip visibility off"
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
